### PR TITLE
Model aliasing

### PR DESCRIPTION
**Pull request by abelsonlive**
# WHAT

This P.R. implements model aliasing RE: https://github.com/fishtown-analytics/dbt/issues/637.

# WHY 

To enable more control over how `dbt` materializes tables.

# HOW 

This P.R. adds an `alias` parameter to the model `config`. If not present, this parameter will default to the `name` of the model (AKA the parsed file path of the model). This implementation is preferable because it maintains the use of the filename in `ref` calls, thereby enabling `dbt` users to have distinct model names, but non-distinct table names. 

The one nasty thing here is that now all materializations reference `model['alias']` when constructing table names (I suspect this might also break some stuff in `dbt-utils`). It may be preferable to add a `table` key to a node so that these references are easier to read.

### CODE REVIEW

File: dbt/context/common.py [hunk 1]
```diff
@@ -45,7 +45,7 @@ def __init__(self, model, adapter, profile):
     def wrap_with_profile_and_model_name(self, fn):
         def wrapped(*args, **kwargs):
             args = (self.profile,) + args
-            kwargs['model_name'] = self.model.get('name')
+            kwargs['model_name'] = self.model.get('alias')
```
Comment (drewbanin) [inline dbt/context/common.py]: I'm not sure we want to do this. The adapters use this `model_name` arg to name their database connection in the pool. I think this line might make it possible for two different models to grab the same connection, which would probably be undesirable.

File: dbt/context/common.py [hunk 2]
```diff
@@ -55,7 +55,7 @@ def type(self):
 
     def commit(self):
         return self.adapter.commit_if_has_connection(
-            self.profile, self.model.get('name'))
+            self.profile, self.model.get('alias'))
```
Comment (drewbanin) [inline dbt/context/common.py]: same here. This arg is used to find the connection in the `default` adapter.

File: dbt/utils.py [hunk 3]
```diff
@@ -318,6 +322,8 @@ def is_blocking_dependency(node):
 def get_materialization(node):
     return node.get('config', {}).get('materialized')
 
+def get_alias(node):
```
Comment (drewbanin) [inline dbt/utils.py]: :+1:

**Comment by drewbanin** [comment]
Couple of things from my first pass:
1. You can make this work for BigQuery by swapping `name` with `alias` [here](https://github.com/fishtown-analytics/dbt/blob/development/dbt/adapters/bigquery.py#L245)
2. The output in the CLI does not reflect the alias for the model. You can find that code [here](https://github.com/fishtown-analytics/dbt/blob/development/dbt/ui/printer.py#L142). There are a couple of other instances where we use `name` that we can replace with `alias`. 

This is so good! It functionally works flawlessly as far as I can tell.

Going to stress test this some more over the weekend, but so far, so good :)

File: dbt/context/common.py (hunk 1 above)
Comment (abelsonlive) [reply dbt/context/common.py]: yeah i think this might be causing bugs on my branch...

**Comment by abelsonlive** [comment]
RE: logging. I was able to fix the lines you mentioned, but i'm still seeing this output:
```
12:44:20 | Concurrency: 1 threads (target='prod')
12:44:20 |
12:44:20 | 1 of 1 START table model foo.model_name [RUN]
12:45:35 | 1 of 1 OK created table model foo.alias [SELECT in 75.23s]
12:45:35 |
12:45:35 | Finished running 1 table models in 75.57s.
```
Where the `START` line prints the model name. Any idea of where that is?

**Comment by drewbanin** [comment]
@abelsonlive ah, yeah, good catch.

The model name is injected in the [describe_node](https://github.com/fishtown-analytics/dbt/blob/development/dbt/node_runners.py#L378) function of these `NodeRunner` classes. I think all of the `describe_node` functions in this file need to be updated to use `'alias'` instead of `'model'`

**Comment by drewbanin** [comment]
@abelsonlive are you blocked on anything here?

**Comment by abelsonlive** [comment]
@drewbanin yeah, it seems like even when I change `name` -> `alias` in the `node_runner.py`, that I still get the same print output.

**Comment by drewbanin** [comment]
@abelsonlive interesting...

In that same `describe_node` function you can do `self.node['config']['alias']`, which correctly returns the alias. Let me try to figure out why `self.node['alias']` isn't updated accordingly. Might be indicative of some bigger oversight

File: dbt/utils.py [hunk 4]
```diff
@@ -39,11 +40,14 @@ def __init__(self, profile, adapter, node, use_temp=False):
         self.node = node
         self.schema = node.get('schema')
         self.name = node.get('name')
+        # set alias, defaults to name
+        self.alias = get_alias(node)
+        self.node['alias'] = get_alias(node)
```
Comment (drewbanin) [inline dbt/utils.py]: oh! Let's not mutate the `node` here!

I think `node['alias']` should be set over [here](https://github.com/fishtown-analytics/dbt/blob/development/dbt/parser.py#L222) instead. The `schema` attribute is set on line 222 and 240 in that file. I think we'll want to do something similar here for `alias`.

This should also fix the issue with the `START` line log output I believe

Comment (abelsonlive) [reply dbt/utils.py]: not mutating the node there caused my initial tests to fail

**Comment by abelsonlive** [comment]
Alright this should be good to go now! Finally got the proper log output. Good look on using `elf.node['config']['alias']`

**Comment by drewbanin** [comment]
@abelsonlive amazing! I think we should probably figure out how to get `self.node['alias']` working before merging this. That's something I can definitely do, as this code looks really good and is pretty much ready to roll notwithstanding the config thing.

Let's try to get this PR in for the 0.9.3 release around a month or so from now. I can spend some time figuring out what's going wrong here right after we drop the 0.9.2 release :)

Thanks for contributing!!

**Pull request closed without merging**